src/service/Offers.ts

```diff
@@ -43,34 +57,72 @@ export const getOfferService = (
       proposalTemplate: { give: giveTemplate, want: wantTemplate },
     } = offer;
 
-    const convertProposals = async paymentProposals => {
+    // Takes give/want entries from dapps and converts them into something
+    // usable in an offer txn. If the entry has an amount, unserialize it. If
+    // the entry has a pursePetname and value, create an amount from those.
+    const convertProposals = async (paymentProposals: GiveOrWantEntries) => {
       const entries = await Promise.all(
         Object.entries(paymentProposals).map(
-          // @ts-expect-error
           async ([kw, { pursePetname, value, amount: serializedAmount }]) => {
-            if (!serializedAmount && !pursePetnameToBrand.get(pursePetname)) {
+            if (
+              !serializedAmount &&
+              !(pursePetname && pursePetnameToBrand.get(pursePetname))
+            ) {
               return [];
             }
 
-            /// TODO: test e2e with dapp inter once feasible.
-            const amount = serializedAmount
-              ? await E(boardIdMarshaller).unserialize(serializedAmount)
-              : AmountMath.make(
-                  pursePetnameToBrand.get(pursePetname),
-                  BigInt(value),
-                );
-
+            let amount =
+              serializedAmount &&
+              (await E(boardIdMarshaller).unserialize(serializedAmount));
+            if (!amount) {
```

Comment:
reading further, I wonder if the it would be clearer to do the checks in sequence.

```js
// first try using serializedAmount
if (serializedAmount) {
  // nested await, but not a hazard in browser
  const amount = await E(boardIdMarshaller).unserialize(serializedAmount);
  return [kw, amount];
}
// next try getting by petname. look up brand just onece.
const brand = pursePetname && pursePetnameToBrand.get(pursePetname);
if (brand) {
  assert(typeof value !== 'undefined');
  const amount = AmountMath.make(brand, BigInt(value));
  return [kw, amount];
}

// can't find it
return [];
```

yeah that's much clear to follow. please simplify with this. (or something like it if I got it wrong)

Reply:
Done thanks for the suggestion